File: h5extract.cpp
/*
 * Usage:
 *      ./h5extract -i INPUT.h5ad -o OUTPUT.npy -c columns.json [-t threads] [-p partitions]
 *
 * columns.json is like
 *
 * {
 *     "columns": [1, 3, 5,... ]
 *     ...
 * }
 *
 */

#include <omp.h>
#include "h5iter.hpp"
#include <iostream>
#include <chrono>
#include <algorithm>
#include <climits>
#include <cmath>
#include <limits>
#include <iomanip>
#include <xtensor/containers/xtensor.hpp>
#include <xtensor/io/xnpy.hpp>
#include <nlohmann/json.hpp>
#include <CLI/CLI.hpp>
#include <fstream>
#include <sstream>


int main(int argc, char *argv[]) {
    std::string input_path;
    std::string output_path;
    std::string columns_path;
    int num_partitions = 200;
    int num_threads = 32;
    int batch_size = 16;

    CLI::App app{"App description"};
    argv = app.ensure_utf8(argv);

    app.add_option("-i,--input", input_path, "Path to the input H5AD file")->required();
    app.add_option("-o,--output", output_path, "Path to the output NPY file")->required();
    app.add_option("-c,--columns", columns_path, "Path to the columns text file")->required();
    app.add_option("-t,--threads", num_threads, "Number of threads to use (default: 32)");
    app.add_option("-p,--partitions", num_partitions, "Number of partitions (default: 100)");

    CLI11_PARSE(app, argc, argv);

    if (num_threads != 0) {
    omp_set_num_threads(num_threads);
    std::cout << "Set OpenMP threads to: " << num_threads << std::endl;
    }

    std::vector<int64_t> columns;
    
    // Load columns from JSON file
    try {
        std::ifstream columns_file(columns_path);
        if (!columns_file.is_open()) {
            std::cerr << "Error: Could not open columns file: " << columns_path << std::endl;
            return 1;
        }
        
        nlohmann::json j;
        columns_file >> j;
        
        if (!j.contains("columns")) {
            std::cerr << "Error: JSON file missing 'columns' field" << std::endl;
            return 1;
        }
        
        columns = j["columns"].get<std::vector<int64_t>>();
    } catch (const std::exception& e) {
        std::cerr << "Error parsing columns file: " << e.what() << std::endl;
        return 1;
    }
    
    if (columns.empty()) {
        std::cerr << "Error: No columns loaded from file: " << columns_path << std::endl;
        return 1;
    }
    
    std::cout << "Loaded " << columns.size() << " columns from " << columns_path << std::endl;
    
    std::map<int64_t, int> column_mapping;
    for (size_t i = 0; i < columns.size(); ++i) {
        column_mapping[columns[i]] = i;
    }


  auto start_time = std::chrono::high_resolution_clock::now();

  h5iter::H5Partitioner partitioner(input_path, num_partitions, batch_size);

  xt::xarray<float> X = xt::zeros<float>({partitioner.rows(), columns.size()});
  
  // Check actual number of threads that will be used
  std::cout << "Starting processing with " << omp_get_max_threads() << " OpenMP threads" << std::endl;
  
  // Global progress tracking
  size_t global_finished_partitions = 0;
  
#pragma omp parallel for schedule(dynamic)
  for (size_t i = 0; i < num_partitions; ++i) {
    auto partition_start_time = std::chrono::high_resolution_clock::now();
    
    for (auto row : partitioner[i].rows_threadsafe()) {
      for (size_t k = 0; k < row.l; k++) {
        int64_t col = row.indices[k];
        float val = row.data[k];
        auto it = column_mapping.find(col);
        if (it == column_mapping.end()) continue;
        X(row.i, it->second) = row.data[k];
      }
    }
    
    #pragma omp critical
    {
      // Update global progress tracking
      global_finished_partitions++;
      auto partition_end_time = std::chrono::high_resolution_clock::now();
      auto partition_duration = std::chrono::duration_cast<std::chrono::milliseconds>(partition_end_time - partition_start_time);
      auto global_elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(partition_end_time - start_time);
      
      // Calculate time estimates
      double avg_time_per_partition = static_cast<double>(global_elapsed.count()) / global_finished_partitions;
      double estimated_total_time = avg_time_per_partition * num_partitions;
      double estimated_remaining_time = estimated_total_time - global_elapsed.count();
      
      int thread_id = omp_get_thread_num();
      std::cout << "Partition " << i << " done (thread " << thread_id << ", " << std::fixed << std::setprecision(1) << (partition_duration.count() / 1000.0) << "s) "
                << "[" << global_finished_partitions << "/" << num_partitions << "] "
                << "Elapsed: " << (global_elapsed.count() / 1000.0) << "s, "
                << "Remaining: " << (estimated_remaining_time / 1000.0) << "s, "
                << "Total: " << (estimated_total_time / 1000.0) << "s" << std::endl;
    }
  }

  xt::dump_npy(output_path, X);
  
  auto end_time = std::chrono::high_resolution_clock::now();
  auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);
  std::cout << "\nExtraction completed in " << (duration.count() / 1000.0) << " seconds" << std::endl;
  std::cout << "Output saved to: " << output_path << std::endl;
  return 0;


}

File: h5extract.py
#!/usr/bin/env python3
"""
Extract columns from H5AD files and save as dense numpy arrays.

Usage:
    python h5extract.py -i INPUT.h5ad -o OUTPUT.npy -c columns.json

columns.json format:
{
    "columns": [1, 3, 5, ...]
}
"""

import argparse
import json
import numpy as np
import scanpy as sc
import sys
from pathlib import Path


def load_columns(columns_path):
    """Load column indices from JSON file."""
    try:
        with open(columns_path, 'r') as f:
            data = json.load(f)
        
        if 'columns' not in data:
            raise ValueError("JSON file missing 'columns' field")
        
        columns = data['columns']
        if not columns:
            raise ValueError("No columns specified in JSON file")
        
        return columns
    except (IOError, json.JSONDecodeError) as e:
        print(f"Error loading columns file: {e}", file=sys.stderr)
        sys.exit(1)
    except ValueError as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


def main():
    parser = argparse.ArgumentParser(description="Extract columns from H5AD files")
    parser.add_argument('-i', '--input', required=True, help='Path to input H5AD file')
    parser.add_argument('-o', '--output', required=True, help='Path to output NPY file')
    parser.add_argument('-c', '--columns', required=True, help='Path to columns JSON file')
    
    args = parser.parse_args()
    
    # Validate input file exists
    if not Path(args.input).exists():
        print(f"Error: Input file not found: {args.input}", file=sys.stderr)
        sys.exit(1)
    
    # Load columns
    columns = load_columns(args.columns)
    print(f"Loaded {len(columns)} columns from {args.columns}")
    
    try:
        # Load H5AD file
        print(f"Loading H5AD file: {args.input}")
        adata = sc.read_h5ad(args.input)
        
        # Extract specified columns
        print(f"Extracting columns from {adata.shape[0]} x {adata.shape[1]} matrix")
        
        # Extract specified columns first (keeping sparse if applicable)
        X_selected = adata.X[:, columns]
        
        # Convert to dense only after extraction
        if hasattr(X_selected, 'toarray'):
            X_selected = X_selected.toarray()
        
        # Save as NPY file
        print(f"Saving extracted matrix ({X_selected.shape[0]} x {X_selected.shape[1]}) to: {args.output}")
        np.save(args.output, X_selected)
        
        print("Extraction completed successfully")
        
    except Exception as e:
        print(f"Error processing H5AD file: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()